# Introduction to Neural Networks with TensorFlow

텐서플로우는 Numpy의 배열과 유사한 자료구조인 tensor를 사용한다.   

tensorflow는 backpropagation을 수행하기 위한 자동 미분 기능을 가지고 있으며 직관적으로 활용할 수 있는 high-level API를 제공한다. 


## Neural Networks

Deep Learning은 1950년 후반에 등장한 artificial neural networks를 기반으로 하고 있다.   
network는 "neurons." 또는 unit 이라고 부르는 단위로 구성된다. 각 unit은 weighted inputs을 입력으로 받는다. 이 weighted inputs 은 선형 결합(linear combination)을 거쳐 하나로 합쳐진다. 그리고 나서 activation function(활성화 함수)를 통과하여 unit의 output 출력된다.

<img src="../assets/simple_neuron.png" width=400px>

수학적으로 표기하면 다음과 같다. 

$$
y = f(h)
$$

여기서, 

$$
h = w_1 x_1 + w_2 x_2 + b = \sum_{i=1}^2 w_i x_i + b
$$

$b = w_0x_0$ ㄹ고 하면,  $h$ 는 두 vector의 dot/inner product 로 표기된다.

$$
h = \begin{bmatrix}
x_0 \, x_1 \, x_2
\end{bmatrix}
\cdot 
\begin{bmatrix}
           w_0 \\
           w_1 \\
           w_2
\end{bmatrix} = w_0 x_0 + w_1 x_1 + w_2 x_2
$$

## Tensors

*tensors*는 행렬을 일반화하여 표현하는 수학적 용어이다.

* vector : 1차원 텐서
* matrix : 2차원 텐서 
* 다차원 행렬 : n차원 텐서 


neural network를 구성하는 기본 data structure가 tensor 이다. 


<img src="../assets/tensor_examples.svg" width=600px>


기본을 익히기 위해 tensor를 사용하여 간단한 신경망 모형을 생성해 보자.

## Import Resources

In [1]:
import warnings
warnings.filterwarnings('ignore')

import numpy as np
import tensorflow as tf

In [2]:
# 파이썬에서 로그 레벨 - 로그는 중요도에 따라 서로 다른 레벨
# DEBUG < INFO < WARNING < ERROR < CRITICAL
# logging 시스템의 레벨은 WARNING
# ERROR로 설정하면 warning은 무시된다.
import logging
logger = tf.get_logger()
logger.setLevel(logging.ERROR)

In [3]:
print('TensorFlow version:', tf.__version__) # 텐서플로우 버전확인

TensorFlow version: 2.3.0


## Single Layer Neural Network

간단한 신경망 모형을 생성하기 위한 랜덤 샘플 데이터를 생성한다.

In [4]:
# Set the random seed so things are reproducible
tf.random.set_seed(7) 

# Create 5 random input features
features = tf.random.normal((1, 5))

# Create random weights for our neural network
weights = tf.random.normal((1, 5))

# Create a random bias term for our neural network
bias = tf.random.normal((1, 1))

print('Features:\n', features)
print('\nWeights:\n', weights)
print('\nBias:\n', bias)

Features:
 tf.Tensor([[ 0.5983449   0.06276207  0.14631742  0.48481876 -0.23572342]], shape=(1, 5), dtype=float32)

Weights:
 tf.Tensor([[-2.2733312  -1.6592104  -0.2633568  -0.80923414  1.0294315 ]], shape=(1, 5), dtype=float32)

Bias:
 tf.Tensor([[1.5749502]], shape=(1, 1), dtype=float32)


실제 데이터를 사용해서 실습하기 앞서 임의의 랜덤 샘플 데이터로 실습을 진행한다.


`tf.Tensor` objects는  data type 과 shape을 가진다. 

* `features = tf.random.normal((1, 5))` `(1, 5)`인 표준 정규분포를 따르는 랜덤 값으로 텐서를 생성한다.

* `weights = tf.random.normal((1, 5))` `(1, 5)`인 표준 정규분포를 따르는 랜덤 값으로 텐서를 생성한다.

* `bias = tf.random.normal((1, 1))` 표준 정규분포를 따르는 임의의 단일한 값을 가지는 텐서를 생성한다. 


텐서플로우의 텐서는 numpy배열과 유사하게 연산이 가능하다. 사용법도 numpy 배열과 유사하다. 위에서 network에서 사용할 텐서를 임의의 값으로 생성하였다.

이제 아래에서 활성화 함수로 sigmoid 함수를 작성해 본다.

In [5]:
def sigmoid_activation(x):
    """ Sigmoid activation function
    
        Arguments
        ---------
        x: tf.Tensor. Must be one of the following types: bfloat16, half, float32, float64, complex64, complex128.
    """
    return 1/(1+tf.exp(-x))

single layer neural network를 생성할 수 있다.

> **실습하기**: 앞서 생성한 `features` ,`weights`, `bias`를 사용하여 신경망 모형을 작성한다. NumPy와 유사하게, TensorFlow 에서 `tf.multiply(a, b)` 를 통해 텐서`a` 와 텐서 `b`의 요소별 곱(multiply)를 구할 수 있다.`tf.reduce_sum(x)` 를 사용하면 tensor `x`의 모든 요소에 대한 합을 구할 수 있다. 활성화 함수로는 앞서 생성한 `sigmoid_activation` function를 사용하도록 한다.

$$
\begin{align}
y &= f(w_1 x_1 + w_2 x_2 + b) \\
y &= f\left(\sum_i w_i x_i +b \right)
\end{align}
$$

In [6]:
## Solution
y = sigmoid_activation(tf.reduce_sum(tf.multiply(features, weights)) + bias)

print('label:\n', y)

label:
 tf.Tensor([[0.3628656]], shape=(1, 1), dtype=float32)


각 텐서의 요소별 곱(multiply) 대신에 matrix multiplication을 사용해도 동일한 결과를 생성할 수 있다.


matrix multiplication를 사용하는 편이 연산의 성능 면에서는 더 좋다. 


`tf.matmul()`를 사용하면 features 와 weights를 matrix multiplication으로 처리 할 수 있다. 대신에 이때 tensor의 shape을 잘 맞추어야만 한다. 맞지 않으면 다음고 같은 에러를 만날 수 있다.

```python
>> tf.matmul(features, weights)

---------------------------------------------------------------------------
InvalidArgumentError                      Traceback (most recent call last)
<ipython-input-7-66a4fe44f20b> in <module>()
      1 
----> 2 y = sigmoid_activation(tf.matmul(features, weights) + bias)

2 frames
/usr/local/lib/python3.6/dist-packages/six.py in raise_from(value, from_value)

InvalidArgumentError: Matrix size-incompatible: In[0]: [1,5], In[1]: [1,5] [Op:MatMul] name: MatMul/
```

위의 에러는 인공 신경망 모형에서 흔히 만날 수 있는 에러인데 matrix multiplication에서 텐서의 shape이 맞지 않을 경우에 해당 에러가 난다.   

이러한 에러는 텐서의 shape(=dim)정보가 맞지 않기 때문에 발생한다.  

다음과 같은 사항을 반드시 주의하자.
- 첫번째 tensor의 column의 갯수와  두번째 tensor의 row 갯수가 서로 일치해야 한다.
- 만약 `features` 가 `weights` 같은 shape `(1, 5)` 라고 하면 `weights` 의 shape를 matrix multiplication 에 맞도록 바꾸어야 한다.

**Note:** `tensor.shape`로 shape를 확인할 수 있다. 


In [7]:
print('Features Shape:', features.shape)
print('Weights Shape:', weights.shape)
print('Bias Shape:', bias.shape)

Features Shape: (1, 5)
Weights Shape: (1, 5)
Bias Shape: (1, 1)


`weights` 의 shape를 matrix multiplication 에 맞도록 바꾸어야 한다.  

`tf.matmul(a,b)` function를 사용해서 matrix multiplication을 수행한다. 이때 shape을 맞추기 위해 `transpose_a`, `transpose_b` arguments 를 사용하면 해당 텐서의 shape을 실제로 변경하지 않고 텐서를 transpose(행과 열을 바꿈)하여 matmul을 수행한다. 이 예제에서는 `transpose_b = True` 를 사용하여 `weights` tensor를 `(1,5)`에서 `(5,1)` 로 변경하여 matmul을 수행하도록 한다.


> **실습** : 위와 동일하게 y를 matrix multiplication로 구현하시오.


In [8]:
## Solution
y = sigmoid_activation(tf.matmul(features, weights, transpose_b = True) + bias)

print('label:\n', y)

label:
 tf.Tensor([[0.3628656]], shape=(1, 1), dtype=float32)


확인해보면, `transpose_b = True` argument를 사용해서 `weights` tensor의 shape을 실제로 변경하지는 않았다. 

In [9]:
print('Weights Shape:', weights.shape)

Weights Shape: (1, 5)


## Multi-Layer Neural Network


개별적인 layer를 여러개 쌓아서 Multi-layer neural network를 구성할 수 있다. layer의 output이 다음 layer의 input이 된다. 여러개의 input units과 output units이 있으므로 weights 는 matrix로 표현된다.

<img src='../assets/multilayer_diagram_weights.png' width=450px>

위 그림에서 맨 밑의 첫 layer는 **input layer** 이다. 중간에 있는 layer는 **hidden layer** 이다.그리고 마지막 layer는 **output layer**이다. 수학적으로 표기하면 다음과 같다.:

$$
\vec{h} = [h_1 \, h_2] = 
\begin{bmatrix}
x_1 \, x_2 \cdots \, x_n
\end{bmatrix}
\cdot 
\begin{bmatrix}
           w_{11} & w_{12} \\
           w_{21} &w_{22} \\
           \vdots &\vdots \\
           w_{n1} &w_{n2}
\end{bmatrix}
$$

network의 output을 수식으로 간단히 표현하면 다음과 같다.

$$
y =  f_2 \! \left(\, f_1 \! \left(\vec{x} \, \mathbf{W_1}\right) \mathbf{W_2} \right)
$$

In [10]:
# Set the random seed so things are reproducible
tf.random.set_seed(7) 

# Create 3 random input features
features = tf.random.normal((1,3))

# Define the size of each layer in our network
n_input = features.shape[1]     # Number of input units, must match number of input features
n_hidden = 2                    # Number of hidden units 
n_output = 1                    # Number of output units

# Create random weights connecting the inputs to the hidden layer
W1 = tf.random.normal((n_input,n_hidden))

# Create random weights connecting the hidden layer to the output layer
W2 = tf.random.normal((n_hidden, n_output))

# Create random bias terms for the hidden and output layers
B1 = tf.random.normal((1,n_hidden))
B2 = tf.random.normal((1, n_output))

> **실습하기:** multi-layer network의 output을 작성해 보시오. weights는 `W1` & `W2`, 그리고  bias는, `B1` & `B2`으로 표현하시오. 

In [11]:

## Solution
h = sigmoid_activation(tf.matmul(features, W1) + B1)
output = sigmoid_activation(tf.matmul(h, W2) + B2)

print(output)

tf.Tensor([[0.10678572]], shape=(1, 1), dtype=float32)


앞선 실습을 제대로 작성하였다면 output은 다음과 같다: `tf.Tensor([[0.10678572]], shape=(1, 1), dtype=float32)`

hidden units의 갯수는 **hyperparameter**이다.  hidden layers 와 hidden units를 많이 할수록 복잡한 문제를 풀수 있는 network capacity가 증가한다.

## NumPy to TensorFlow and Back

tensor는 numpy의 `ndarrays` 와 호환이 된다. `tf.convert_to_tensor()`를 사용하면 NumPy ndarray를 tensor로 변환할 수 있다. `.numpy()`  메소드를 사용하면 tensor를 NumPy array로 변환할 수 있다.

In [12]:
# Set the random seed so things are reproducible
tf.random.set_seed(7) 

a = np.random.rand(4,3)

print(type(a))
print(a)

<class 'numpy.ndarray'>
[[0.05555019 0.88001085 0.37259558]
 [0.99069457 0.73259237 0.18435513]
 [0.7141118  0.78794829 0.37554902]
 [0.56149971 0.33285178 0.7345163 ]]


In [13]:
b = tf.convert_to_tensor(a)

print(b)

tf.Tensor(
[[0.05555019 0.88001085 0.37259558]
 [0.99069457 0.73259237 0.18435513]
 [0.7141118  0.78794829 0.37554902]
 [0.56149971 0.33285178 0.7345163 ]], shape=(4, 3), dtype=float64)


In [14]:
c = b.numpy()

print(type(c))
print(c)

<class 'numpy.ndarray'>
[[0.05555019 0.88001085 0.37259558]
 [0.99069457 0.73259237 0.18435513]
 [0.7141118  0.78794829 0.37554902]
 [0.56149971 0.33285178 0.7345163 ]]


tensor 의 value를 변경하여도 ndarray의 value는 변경되지 않는다.

In [15]:
# Multiply TensorFlow Tensor by 40
b = b * 40

print(b)

tf.Tensor(
[[ 2.22200767 35.20043392 14.903823  ]
 [39.62778298 29.30369474  7.37420506]
 [28.56447189 31.51793176 15.02196063]
 [22.45998845 13.31407135 29.38065216]], shape=(4, 3), dtype=float64)


In [16]:
# NumPy array stays the same
a

array([[0.05555019, 0.88001085, 0.37259558],
       [0.99069457, 0.73259237, 0.18435513],
       [0.7141118 , 0.78794829, 0.37554902],
       [0.56149971, 0.33285178, 0.7345163 ]])

In [17]:
# NumPy array stays the same
c

array([[0.05555019, 0.88001085, 0.37259558],
       [0.99069457, 0.73259237, 0.18435513],
       [0.7141118 , 0.78794829, 0.37554902],
       [0.56149971, 0.33285178, 0.7345163 ]])

In [18]:
# Add 1 to NumPy ndarray
a = a + 1

print(a)

[[1.05555019 1.88001085 1.37259558]
 [1.99069457 1.73259237 1.18435513]
 [1.7141118  1.78794829 1.37554902]
 [1.56149971 1.33285178 1.7345163 ]]


In [19]:
# Tensor stays the same
print(b)

tf.Tensor(
[[ 2.22200767 35.20043392 14.903823  ]
 [39.62778298 29.30369474  7.37420506]
 [28.56447189 31.51793176 15.02196063]
 [22.45998845 13.31407135 29.38065216]], shape=(4, 3), dtype=float64)


In [20]:
# NumPy array stays the same
c

array([[0.05555019, 0.88001085, 0.37259558],
       [0.99069457, 0.73259237, 0.18435513],
       [0.7141118 , 0.78794829, 0.37554902],
       [0.56149971, 0.33285178, 0.7345163 ]])